### PR TITLE
Add SPARC architecture

### PR DESCRIPTION
**Pull request by yashi**
This PR is to add SPARC architecture to Zephyr RTOS.  This PR is only for QEMU target.

It's my first PR to add new architecture to Zephyr.  Please let me know if there is any procedure I should follow to.

This PR needs #16950 to be merged before.

Hoping to be merged for version 2.0.

### CODE REVIEW

File: cmake/compiler/gcc/target.cmake
Comment (SebastianBoe) [inline cmake/compiler/gcc/target.cmake]: indentation

File: cmake/compiler/gcc/target.cmake
Comment (SebastianBoe) [inline cmake/compiler/gcc/target.cmake]: Don't add -mflat twice.

File: cmake/compiler/gcc/target.cmake
Comment (SebastianBoe) [inline cmake/compiler/gcc/target.cmake]: Either add an explanation to why this needs to be different with SPARC as a comment in the code, or preferably, align with the other architecture's and avoid this patch.

File: drivers/timer/Kconfig
Comment (SebastianBoe) [inline drivers/timer/Kconfig]: It is best-practice to omit default n.

File: soc/sparc/at697/CMakeLists.txt
Comment (SebastianBoe) [inline soc/sparc/at697/CMakeLists.txt]: Find a way to avoid specifying -mcpu=leon multiple times.

File: arch/Kconfig
Comment (andrewboie) [inline arch/Kconfig]: why? shouldn't you be using GCC builtins?

File: arch/sparc/core/cpu_idle.c
Comment (andrewboie) [inline arch/sparc/core/cpu_idle.c]: this function will immediately return - does SPARC have a "wait for interrupt" instruction because that's what this function needs to do

File: arch/sparc/core/entry.S
Comment (andrewboie) [inline arch/sparc/core/entry.S]: you need to namespace all your private variables and functions to this arch port with z_, preferably z_sparc_ for clarity

**Review by galak — CHANGES_REQUESTED** [review]
This needs an update to sdk-ng to build a toolchain for SPARC.

File: arch/sparc/core/entry.S
Comment (yashi) [reply arch/sparc/core/entry.S]: done.

File: drivers/timer/Kconfig
Comment (yashi) [reply drivers/timer/Kconfig]: done

File: cmake/compiler/gcc/target.cmake
Comment (yashi) [reply cmake/compiler/gcc/target.cmake]: fixed.

File: arch/Kconfig
Comment (yashi) [reply arch/Kconfig]: Current build doesn't link against libatomic and failed with the following error:
```
include/sys/atomic.h:223: undefined reference to `__atomic_fetch_or_4'
include/sys/atomic.h:270: undefined reference to `__atomic_fetch_and_4'
```

In fact, many arch currently selects pure C version of it.   Would it be OK with you if I say I'll create an issue to convert to builtins once this PR is merged?

Comment (andrewboie) [reply arch/Kconfig]: toolchain may not implement them, we had this issue with Nios II. fine to use C for now

File: arch/sparc/core/cpu_idle.c
Comment (yashi) [reply arch/sparc/core/cpu_idle.c]: Unfortunately it doesn't have the instruction.  Instead, it's a implementation dependent.  Some LEON3 has it in Ancillary State Register:

> 2.2.13 Power-down
> The processor can be configured to include a power-down feature to minimize power consumption during idle periods. The power-down mode is entered by performing a WRASR instruction to %asr19:
> ```
> wr %g0, %asr19
> ```
> During power-down, the pipeline is halted until the next interrupt occurs. Signals inside the processor pipeline and caches are then static, reducing power consumption from dynamic switching.

But this PR is for LEON **2**, which doesn't have it, neither.

Comment (andrewboie) [reply arch/sparc/core/cpu_idle.c]: OK

**Review by tyrantlizard77 — DISMISSED** [review]
LGTM once SDK and CI issues resolved

**Comment by yashi** [comment]
Thanks.  Regarding to the CI issues, I mean the warnings and errors `checkpatch`, I think I've fixed I can.  The remains are:

- WARNING:NEW_TYPEDEFS: do not add new typedefs
  - I think I need typedefs because these are used by `scripts/gen_offset_header.py`
- ERROR:MULTISTATEMENT_MACRO_USE_DO_WHILE: Macros with multiple statements should be enclosed in a do - while loop
  - I don't think we can wrap linker directives with do-while.
- typedef off_t
  - need to have it.

What should I do about them?

**Comment by andrewboie** [comment]
Ah right, checkpatch errors don't hold up merge, whoever merges will take them into account. Shippable is happy, so if we can get a few more reviews in the next few days we should be able to make for 2.0

Is this arch currently being built by CI?

**Comment by yashi** [comment]
No.  Because the SDK don't have a SPARC compiler, I don't think the CI is building it. I'll take a look at sdk-ng after I create a separate issue about libgcc and compiler flags @SebastianBoe commented.

**Comment by galak** [comment]
> No. Because the SDK don't have a SPARC compiler, I don't think the CI is building it. I'll take a look at sdk-ng after I create a separate issue about libgcc and compiler flags @SebastianBoe commented.

What toolsuite are you using to develop this?

**Comment by yashi** [comment]
I've been using GCC from https://www.gaisler.com/index.php/downloads/compilers .
I've tried `gcc-8-sparc64-linux-gnu` from Debian but haven't successfully built; always get `SPARC32PLUS, V8+ Required` set in the final binary, even with `-mno-v8plus`.

**Comment by yashi** [comment]
Just in case anyone is interested in.  A diff of `gcc -Q --help=target`:
```diff
--- /tmp/debian.txt	2019-08-07 19:45:25.033387977 +0900
+++ /tmp/bcc.txt	2019-08-07 19:45:30.637429500 +0900
@@ -1,40 +1,36 @@
 
-  -m32                        		[disabled]
-  -m64                        		[enabled]
+  -m32                        		[enabled]
+  -m64                        		[disabled]
   -mapp-regs                  		[enabled]
-  -mbionic                    		[disabled]
   -mcbcond                    		[disabled]
   -mcmodel=                   		
-  -mcpu=                      		v9
+  -mcpu=                      		v7
   -mdebug=                    		
   -mfaster-structs            		[disabled]
   -mfix-at697f                		[disabled]
+  -mfix-b2bst                 		[disabled]
   -mfix-gr712rc               		[disabled]
+  -mfix-tn0013                		[disabled]
   -mfix-ut699                 		[disabled]
   -mfix-ut700                 		[disabled]
   -mflat                      		[disabled]
   -mfmaf                      		[disabled]
   -mfpu                       		[enabled]
-  -mfsmuld                    		[enabled]
-  -mglibc                     		[enabled]
+  -mfsmuld                    		[disabled]
   -mhard-float                		[enabled]
   -mhard-quad-float           		[disabled]
-  -mlong-double-128           		[enabled]
-  -mlong-double-64            		[disabled]
-  -mlra                       		[enabled]
-  -mmemory-model=             		tso
-  -mmusl                      		[disabled]
+  -mlra                       		[disabled]
+  -mmemory-model=             		default
   -mpopc                      		[disabled]
-  -mptr32                     		[disabled]
-  -mptr64                     		[enabled]
+  -mptr32                     		[enabled]
+  -mptr64                     		[disabled]
   -mrelax                     		[disabled]
   -msoft-float                		[disabled]
   -msoft-quad-float           		[enabled]
-  -mstack-bias                		[enabled]
+  -mstack-bias                		[disabled]
   -mstd-struct-return         		[disabled]
   -msubxc                     		[disabled]
-  -mtune=                     		v9
-  -muclibc                    		[disabled]
+  -mtune=                     		v7
   -munaligned-doubles         		[disabled]
   -muser-mode                 		[enabled]
   -mv8plus                    		[disabled]
```

**Comment by galak** [comment]
> I've been using GCC from [gaisler.com/index.php/downloads/compilers](https://www.gaisler.com/index.php/downloads/compilers) .
> I've tried `gcc-8-sparc64-linux-gnu` from Debian but haven't successfully built; always get `SPARC32PLUS, V8+ Required` set in the final binary, even with `-mno-v8plus`.

and which qemu?

**Comment by yashi** [comment]
For QEMU, we've been using https://github.com/AdaCore/qemu with a few fixes.

**Comment by galak** [comment]
I know you hoped to get this into 2.0.0, but the cut off for that is this Friday.  Since we don't get have the SDK support needed for this it seems highly unlikely we'll be able to get this in for 2.0.0.  However I don't see any reason that we can't get this in for 2.1.0

**Comment by yashi** [comment]
@galak, I agree.  I thought it'd be a nice addition (a new arch) for a major version bump.  But, it's just that.  No rush from me to squeeze it into 2.0.  Thank you all, as always, for the kind support.  I'll finish other things (preparing sdk and the libgcc issue) first and then come back to this after the 2.0 release.

File: arch/sparc/core/cpu_idle.c
Comment (andrewboie) [reply arch/sparc/core/cpu_idle.c]: can you declare this __weak so it may be overridden in board or SOC configuration?

File: arch/sparc/core/irq_offload.c
Comment (andrewboie) [inline arch/sparc/core/irq_offload.c]: don't need this anymore once #20397 lands (very soon)

**Review by tyrantlizard77 — APPROVED** [review]
some minor suggestions, otherwise looks great
please address @SebastianBoe 's comments

**Comment by yashi** [comment]
Took me a while to came back to this.  I've rebased to the current master.   Addressed a few places pointed by reviews.  I haven't fixed points where @SebastianBoe pointed out, just yet.  I'll work on them next.  This push is to see how everything else goes.

File: arch/sparc/core/irq_offload.c
Comment (yashi) [reply arch/sparc/core/irq_offload.c]: Removed sched lock.

File: arch/sparc/core/cpu_idle.c
Comment (yashi) [reply arch/sparc/core/cpu_idle.c]: Added __weak to `arch_cpu_idle()` and comment about it in the commit log.

**Comment by andrewboie** [comment]
Where are we at with getting a SPARC toolchain/emulator in the SDK?

**Comment by stephanosio** [comment]
> Where are we at with getting a SPARC toolchain/emulator in the SDK?

Both toolchain and emulator (qemu) were added in Zephyr SDK 0.11.0 Alpha 7, which will be available in the CI image once #20625 is merged.

**Comment by yashi** [comment]
> This needs an update to sdk-ng to build a toolchain for SPARC.

It has been merged.
https://github.com/zephyrproject-rtos/sdk-ng/pull/148

**Comment by yashi** [comment]
Forgot to push LEON2 patches for QEMU.  The current patch is just to enable SPARC arch.  I'll address it in sdk-ng.

File: cmake/compiler/gcc/target.cmake
Comment (yashi) [reply cmake/compiler/gcc/target.cmake]: We'll postpone this change after sparc is merged.  Because, for now, we are targeting GCC in SDK, we don't need this.   I'll later open another issue just for this.

File: cmake/compiler/gcc/target.cmake
Comment (yashi) [reply cmake/compiler/gcc/target.cmake]: I've moved `-mflat` from `arch/sparc/CMakeLists.txt` and `cmake/compiler/gcc/target.cmake` to `cmake/compiler/gcc/target_sparc.cmake` like arm and x86.

File: soc/sparc/at697/CMakeLists.txt
Comment (yashi) [reply soc/sparc/at697/CMakeLists.txt]: In the current version, we only have "-mcpu=leon -mtune=leon" only in this file.

**Comment by yashi** [comment]
I've addressed everything reviewers pointed out.  Thanks all.  Let me know if there still is anything I can fix.

**Comment by yashi** [comment]
Rebased on v2.1.

**Comment by andrewboie** [comment]
Needs a rebase.

@galak the new SDK is in place, can you re-visit your review?

**Comment by yashi** [comment]
I've rebased to the current master.  It compiles and runs `samples/philosophers` just fine.

Unfortunately, many tests by `sanitycheck` fail.  What is the best way to disable unsupported test in `boards/*/*/*.yaml`?  I was gonna use `ignore_tags:` but it's kind of tedious to list them all.  Is there better way?

**Comment by yashi** [comment]
Hmmm it seems it's qemu.  I'll take a look.

**Pull request closed without merging**
